### PR TITLE
Allow rate limiting configuration overrides from external sources

### PR DESCRIPTION
**Pull request by neoaggelos**
#### Summary

Closes https://github.com/TheThingsIndustries/lorawan-stack/issues/2676

#### Changes

- Move rate limiting configuration to `pkg/config` package.
- Support loading profiles from external sources (directory, URL, bucket)
- External configuration has higher priority, allowing operators to maintain a core rate limiting configuration, and easily apply overrides as needed.

#### Testing

- [X] locally
- [x] unit tests (TODO)

#### Notes for Reviewers

Draft until discussion in https://github.com/TheThingsIndustries/lorawan-stack/issues/2676 is resolved. Targetting 3.13.0 since this PR introduces configuration changes.

#### Checklist

- [X] Scope: The referenced issue is addressed, there are no unrelated changes.
- [X] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [X] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: pkg/config/shared.go [hunk 1]
```diff
@@ -450,3 +449,48 @@ type MQTTConfigProviderFunc func(context.Context) (*MQTT, error)
...
+	MaxBurst     uint     `name:"max-burst" yaml:"max-burst" description:"Maximum rate allowed for short bursts"`
+	Associations []string `name:"associations" description:"List of classes to apply this profile on"`
+}
+
+// RateLimitingMemory represents configuration for the in-memory rate limiting store.
+type RateLimitingMemory struct {
+	MaxSize uint `name:"max-size" description:"Maximum store size for the rate limiter"`
+}
+
+// RateLimitingConfig represents configuration for rate limiting.
```
Comment (KrishnaIyer) [inline pkg/config/shared.go]: `RateLimiting`

File: pkg/ratelimit/config.go [hunk 2]
```diff
@@ -19,49 +19,51 @@ import (
...
 		return defaultLimiter, nil
 	}
 
 	l := &muxRateLimiter{
 		defaultLimiter: defaultLimiter,
-		limiters:       make(map[string]Interface, len(c.Profiles)),
+		limiters:       make(map[string]Interface, len(profiles)),
 	}
-	for _, profile := range c.Profiles {
+	for _, profile := range profiles {
```
Comment (KrishnaIyer) [inline pkg/ratelimit/config.go:62]: So you merge the profiles here from the various files?

**Review by KrishnaIyer — COMMENTED** [review]
Implementation looks good to me. Haven't checked all the details since I didn't review the original PR.

1. Does this break existing configuration?
2. > (embedded configs would take priority).

I think the external one should take priority since we would like to explicitly override embedded config.

File: pkg/ratelimit/config.go (hunk 2 above)
Comment (neoaggelos) [reply pkg/ratelimit/config.go:62]: Yes. Profiles from external configuration are set last, and so will take priority.

**Review by htdvisser — APPROVED** [review]
LGTM

**Review by KrishnaIyer — APPROVED** [review]
Good stuff. LGTM

**Review by adriansmares — APPROVED** [review]

**Pull request merged**